### PR TITLE
[DAG] visitCTPOP - if only the upper half of the ctpop operand is zero then see if its profitable to only count the lower half.

### PR DESCRIPTION
**Pull request by RKSimon**

### CODE REVIEW

**Comment by phoebewang** [comment]
Do we still need https://github.com/llvm/llvm-project/commit/275729ae06d568e9589392c142a416fb8c2bb1a8 with this change?

**Comment by RKSimon** [comment]
> Do we still need [275729a](https://github.com/llvm/llvm-project/commit/275729ae06d568e9589392c142a416fb8c2bb1a8) with this change?

Yes, I've also further extended it with 3a758076f54d521d20e32856a62c7d24803ee9e8 - I did consider doing something similar here in DAGCombine if shifting down the bits would allow a smaller CTPOP instruction, but decided against it - it might still be useful in legalization though?

**Comment by phoebewang** [comment]
> > Do we still need [275729a](https://github.com/llvm/llvm-project/commit/275729ae06d568e9589392c142a416fb8c2bb1a8) with this change?
> 
> Yes, I've also further extended it with [3a75807](https://github.com/llvm/llvm-project/commit/3a758076f54d521d20e32856a62c7d24803ee9e8) - I did consider doing something similar here in DAGCombine if shifting down the bits would allow a smaller CTPOP instruction, but decided against it - it might still be useful in legalization though?

Do you mean with https://github.com/llvm/llvm-project/commit/3a758076f54d521d20e32856a62c7d24803ee9e8, we still need to customize i16/32/64 so that doesn't matter?

**Comment by RKSimon** [comment]
I mean with 3a758076f54d521d20e32856a62c7d24803ee9e8 we can now search all integer ctpop ops for "shifted active bits" cases that allow us to identify when we can make use of more of fast implementations listed in #79823

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -11142,11 +11142,29 @@ SDValue DAGCombiner::visitCTTZ_ZERO_UNDEF(SDNode *N) {
...
   if (SDValue C = DAG.FoldConstantArithmetic(ISD::CTPOP, DL, VT, {N0}))
     return C;
+
+  // If the upper bits are known to be zero, then see if its profitable to
+  // only count the lower bits.
+  if (VT.isScalarInteger() && NumBits > 8 && (NumBits & 1) == 0) {
+    EVT HalfVT = EVT::getIntegerVT(*DAG.getContext(), NumBits / 2);
+    if (TLI.isTruncateFree(N0, HalfVT) && TLI.isZExtFree(HalfVT, VT) &&
+        TLI.isTypeDesirableForOp(ISD::CTPOP, HalfVT) &&
+        hasOperation(ISD::CTPOP, HalfVT)) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:11158]: Check these two conditions first?

**Review by arsenm — APPROVED** [review]

**Pull request merged**